### ISSUE TITLE
Do we actually need hash tagging?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#2937]
To be clear, I don't mean sha256t.

I've just realized after the changes to remove `ThirtyTwoByteHash` trait and only impl `Into<Message> for Sighash` we don't need to worry about whether non-sighash hashes are a hash (unless I'm missing something else). And if sighashes are the only things that need to ensure it, maybe the whole type should be protected?

Basically, I think we should be able to not have `from_byte_array` on sighashes (we need to modify the macro) and then just explicitly have `const fn dangerous_from_byte_array_assume_hashed(x: [u8; 32]) -> Self`.

Not having to deal with tags and simply delegating the responsibility to individual type is simpler and could get us to stabilize faster. Also we can always add support for non-raw hashes later by adding a flag to the macro.

**Comment by apoelstra** [comment]
To further clarify: by "hash tagging" you mean adding a type tag to indicate whether or not a preimage is known.

It's an interesting point -- we basically have this in that the *`SigHash` types are "known preimage" and the rest of them aren't. And with the sighash types we have already restricted the API so that it's unnatural to directly serialize/deserialize them ... and maybe you're right that even the "unnatural" way should be more clearly marked as dangerous.

I think that just tacking `assume_hashed` onto the end of `from_byte_array`, and mentioning this in the docs of `to_byte_array`, is enough. We don't need to say `dangerous`. The word "assume" in a method name already means "dangerous" IMO.

TL;DR I think you're right.

**Comment by tcharding** [comment]
Am I correct in thinking this issue is a counter argument to https://github.com/rust-bitcoin/rust-bitcoin/issues/1425?

And is the suggestion to change the name of `from_byte_array` to  `from_byte_array_assume_hashed` for some subset of hash types?

I'm not understanding the suggested difference between a "known pre-image" sighash type and an "unknown pre-image"? For example a `Txid` unknown or known?

**Comment by apoelstra** [comment]
> Am I correct in thinking this issue is a counter argument to #1425?

Yep.

> For example a `Txid` unknown or known?

A `Txid` you parse from an outpoint is unknown. A `Txid` you compute by hashing a transaction is known. So our `Txid` type would have to be split into two -- `Txid<UnknownPreimage>` and `Txid<KnownPreimage>` where the latter let you do more stuff.

The point of this issue is that the "more stuff you can do" if you know a hash preimage is frequently the empty set. It lets you sign hashes (but you would never directly sign a `Txid`, only a `TaprootSighash` or whatever); it gives you assurance that txids won't collide (but this isn't even meaningful if you don't know the preimages, so it's fine); it lets you say "this represents a real" transaction (but the definition of "real" here is "it's a series of bytes" so this is a useless concept).

IOW, for non-sighashes the "unknown preimage" type is really the only useful one. And conversely, for sighashes, the "known preimage" type is the only useful one (and we want to discourage users from deserializing these objects because doing so is usually unsafe).

> And is the suggestion to change the name of `from_byte_array` to `from_byte_array_assume_hashed` for some subset of hash types?

Yep -- specifically hash types which implement `Into<Message>`. (There is an "obviously better" way to accomplish this with a custom trait that has the methods alongside a conversion to `Message`. We do not want to do this because it will make it very hard/complicated to avoid a circular dependency between `hashes` and `secp256k1` and the custom trait would really only be useful for rust-secp itself.)

So I think we can close #1425.

**Comment by apoelstra** [comment]
Closing this issue. We have landed on "known-preimage hashes are exactly the ones that impl `Into<Message>`".
